File: comm.zig
//! comm.zig – Zig port of comm.c.
//! IPC pipes between the main swaylock process and the PAM child.

const std = @import("std");
const types = @import("types.zig");

const log = @import("log.zig");
const password_buffer = @import("password_buffer.zig");

/// Maximum payload size accepted from the pipe (1 MiB).
const comm_max_payload: usize = 1 << 20;

// comm_fds[0]: main→child  (main writes [1], child reads [0])
// comm_fds[1]: child→main  (child writes [1], main reads [0])
var comm_fds: [2][2]i32 = .{ .{ -1, -1 }, .{ -1, -1 } };

fn slogErrno(
    verbosity: log.LogImportance,
    src: std.builtin.SourceLocation,
    comptime msg: []const u8,
    err: anyerror,
) void {
    log.slog(verbosity, src, msg ++ ": {s}", .{@errorName(err)});
}

fn readFull(fd: i32, dst: []u8) isize {
    var offset: usize = 0;
    while (offset < dst.len) {
        const n = std.posix.read(fd, dst[offset..]) catch |err| {
            if (err == error.Interrupted) continue;
            slogErrno(log.LogImportance.err, @src(), "read() failed", err);
            return -1;
        };
        if (n == 0) {
            if (offset == 0) return 0;
            log.slog(
                log.LogImportance.err,
                @src(),
                "read() failed: unexpected EOF",
                .{},
            );
            return -1;
        }
        offset += n;
    }
    return @intCast(offset);
}

fn writeFull(fd: i32, src: []const u8) bool {
    var offset: usize = 0;
    while (offset < src.len) {
        const n = std.posix.write(fd, src[offset..]) catch |err| {
            if (err == error.Interrupted) continue;
            slogErrno(log.LogImportance.err, @src(), "write() failed", err);
            return false;
        };
        if (n == 0) {
            log.slog(log.LogImportance.err, @src(), "write() returned 0", .{});
            return false;
        }
        offset += n;
    }
    return true;
}

fn loadLe32(b: *const [4]u8) u32 {
    return @as(u32, b[0]) |
        (@as(u32, b[1]) << 8) |
        (@as(u32, b[2]) << 16) |
        (@as(u32, b[3]) << 24);
}

fn storeLe32(b: *[4]u8, v: u32) void {
    b[0] = @truncate(v);
    b[1] = @truncate(v >> 8);
    b[2] = @truncate(v >> 16);
    b[3] = @truncate(v >> 24);
}

fn commRead(
    fd: i32,
    payload: *?[*]u8,
    len: *usize,
) i32 {
    var msg_type: u8 = undefined;
    var n = readFull(fd, std.mem.asBytes(&msg_type));
    if (n <= 0) {
        payload.* = null;
        return @intCast(n);
    }
    var plen_buf: [4]u8 = undefined;
    n = readFull(fd, &plen_buf);
    if (n <= 0) {
        payload.* = null;
        return -1;
    }
    const plen: usize = loadLe32(&plen_buf);
    if (plen > comm_max_payload) {
        log.slog(
            log.LogImportance.err,
            @src(),
            "comm_read: payload too large: {d}",
            .{plen},
        );
        payload.* = null;
        return -1;
    }
    var buf: ?[*]u8 = null;
    if (plen > 0) {
        buf = @ptrCast(std.c.malloc(plen + 1));
        if (buf == null) {
            log.slog(log.LogImportance.err, @src(), "allocation failed", .{});
            payload.* = null;
            return -1;
        }
        n = readFull(fd, buf.?[0..plen]);
        if (n <= 0) {
            std.c.free(@ptrCast(buf));
            payload.* = null;
            return -1;
        }
        buf.?[plen] = 0;
    }
    payload.* = buf;
    len.* = plen;
    return @intCast(msg_type);
}

fn commWrite(
    fd: i32,
    msg_type: u8,
    payload: []const u8,
) bool {
    if (!writeFull(fd, std.mem.asBytes(&msg_type))) return false;
    var plen_buf: [4]u8 = undefined;
    storeLe32(&plen_buf, @intCast(payload.len));
    if (!writeFull(fd, &plen_buf)) return false;
    if (payload.len > 0) {
        if (!writeFull(fd, payload)) return false;
    }
    return true;
}

/// Returns the fd the child reads incoming messages from.
pub fn getCommChildFd() i32 {
    return comm_fds[0][0];
}

/// Reads a message from the child-facing pipe.
pub fn commChildRead(payload: *?[*]u8, len: *usize) i32 {
    return commRead(comm_fds[0][0], payload, len);
}

/// Writes a message to the child-facing pipe.
pub fn commChildWrite(
    msg_type: u8,
    payload: []const u8,
) bool {
    return commWrite(comm_fds[1][1], msg_type, payload);
}

/// Reads a message from the main-facing pipe.
pub fn commMainRead(payload: *?[*]u8, len: *usize) i32 {
    return commRead(comm_fds[1][0], payload, len);
}

/// Writes a message to the main-facing pipe.
pub fn commMainWrite(
    msg_type: u8,
    payload: []const u8,
) bool {
    return commWrite(comm_fds[0][1], msg_type, payload);
}

/// Returns the fd to poll for messages from the child.
pub fn getCommReplyFd() i32 {
    return comm_fds[1][0];
}

/// Clears and sends the password buffer as a COMM_MSG_PASSWORD frame.
/// The password buffer is always cleared before returning.
pub fn writeCommPassword(pw: *types.SwaylockPassword) bool {
    const size: usize = @intCast(pw.len + 1);
    const copy = password_buffer.create(size);
    if (copy == null) {
        password_buffer.clear(pw);
        return false;
    }
    @memcpy(copy.?[0..size], pw.buffer.?[0..size]);
    password_buffer.clear(pw);
    const ok = commWrite(
        comm_fds[0][1],
        types.CommMsg.password,
        copy.?[0..size],
    );
    password_buffer.destroy(copy.?[0..size]);
    return ok;
}

/// Spawns the comm child process.
pub fn spawnCommChild(child_fn: *const fn () void) bool {
    const fds0 = std.posix.pipe() catch |err| {
        slogErrno(log.LogImportance.err, @src(), "failed to create pipe", err);
        return false;
    };
    comm_fds[0] = fds0;
    const fds1 = std.posix.pipe() catch |err| {
        slogErrno(log.LogImportance.err, @src(), "failed to create pipe", err);
        return false;
    };
    comm_fds[1] = fds1;
    const child = std.posix.fork() catch |err| {
        slogErrno(log.LogImportance.err, @src(), "failed to fork", err);
        return false;
    };
    if (child == 0) {
        const act = std.posix.Sigaction{
            .handler = .{ .handler = std.posix.SIG.IGN },
            .mask = std.posix.empty_sigset,
            .flags = 0,
        };
        std.posix.sigaction(
            std.posix.SIG.USR1,
            &act,
            null,
        );
        std.posix.close(comm_fds[0][1]);
        std.posix.close(comm_fds[1][0]);
        // Redirect stdin and stdout to /dev/null so the PAM
        // module cannot fall back to prompting on the terminal
        // if the main process exits or authd is unavailable.
        if (std.posix.open(
            "/dev/null",
            .{ .ACCMODE = .RDWR },
            0,
        )) |devnull| {
            std.posix.dup2(devnull, 0) catch {};
            std.posix.dup2(devnull, 1) catch {};
            if (devnull > 1) std.posix.close(devnull);
        } else |_| {}
        child_fn();
        // child_fn calls exit(); unreachable
        unreachable;
    }
    std.posix.close(comm_fds[0][0]);
    std.posix.close(comm_fds[1][1]);
    return true;
}

File: comm.c
#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#include "comm.h"
#include "log.h"
#include "password-buffer.h"
#include "swaylock.h"

/*
 * comm[0]: main→child  (main writes [1], child reads [0])
 * comm[1]: child→main  (child writes [1], main reads [0])
 */
static int comm[2][2] = {{-1, -1}, {-1, -1}};

/*
 * Returns the fd the child reads incoming messages from.
 * Not declared in comm.h; pam.c references it via an extern declaration.
 */
int get_comm_child_fd(void) {
	return comm[0][0];
}

static ssize_t read_full(int fd, void *dst, size_t size) {
	char *buf = dst;
	size_t offset = 0;
	while (offset < size) {
		ssize_t n = read(fd, &buf[offset], size - offset);
		if (n < 0) {
			if (errno == EINTR) {
				continue;
			}
			swaylock_log_errno(LOG_ERROR, "read() failed");
			return -1;
		} else if (n == 0) {
			if (offset == 0) {
				return 0;
			}
			swaylock_log(LOG_ERROR,
				"read() failed: unexpected EOF");
			return -1;
		}
		offset += n;
	}
	return (ssize_t)offset;
}

static bool write_full(int fd, const void *src, size_t size) {
	const char *buf = src;
	size_t offset = 0;
	while (offset < size) {
		ssize_t n = write(fd, &buf[offset], size - offset);
		if (n <= 0) {
			assert(n != 0);
			if (errno == EINTR) {
				continue;
			}
			swaylock_log_errno(LOG_ERROR, "write() failed");
			return false;
		}
		offset += n;
	}
	return true;
}

/* Read a 4-byte little-endian uint32 from a byte buffer. */
static uint32_t load_le32(const uint8_t *b) {
	return (uint32_t)b[0]
		| ((uint32_t)b[1] << 8)
		| ((uint32_t)b[2] << 16)
		| ((uint32_t)b[3] << 24);
}

/* Write a 4-byte little-endian uint32 into a byte buffer. */
static void store_le32(uint8_t *b, uint32_t v) {
	b[0] = (uint8_t)(v & 0xff);
	b[1] = (uint8_t)((v >> 8)  & 0xff);
	b[2] = (uint8_t)((v >> 16) & 0xff);
	b[3] = (uint8_t)((v >> 24) & 0xff);
}

/*
 * Frame layout: uint8_t type | uint32_t payload_len (LE) | payload[plen]
 *
 * Returns message type (>0) on success, 0 on EOF, -1 on error.
 * On success *payload is malloc'd (caller must free) and *len is set.
 * On EOF/error *payload is set to NULL.
 */
static int comm_read(int fd, char **payload, size_t *len) {
	uint8_t type;
	ssize_t n = read_full(fd, &type, sizeof(type));
	if (n <= 0) {
		*payload = NULL;
		return (int)n;
	}

	uint8_t plen_buf[4];
	n = read_full(fd, plen_buf, sizeof(plen_buf));
	if (n <= 0) {
		*payload = NULL;
		return -1;
	}

	uint32_t plen = load_le32(plen_buf);
	char *buf = NULL;
	if (plen > 0) {
		buf = malloc(plen);
		if (!buf) {
			swaylock_log(LOG_ERROR, "allocation failed");
			*payload = NULL;
			return -1;
		}
		n = read_full(fd, buf, plen);
		if (n <= 0) {
			free(buf);
			*payload = NULL;
			return -1;
		}
	}

	*payload = buf;
	*len = plen;
	return (int)type;
}

static bool comm_write(
	int fd, uint8_t type, const char *payload, size_t len)
{
	if (!write_full(fd, &type, sizeof(type))) {
		return false;
	}
	uint8_t plen_buf[4];
	store_le32(plen_buf, (uint32_t)len);
	if (!write_full(fd, plen_buf, sizeof(plen_buf))) {
		return false;
	}
	if (len > 0 && payload) {
		if (!write_full(fd, payload, len)) {
			return false;
		}
	}
	return true;
}

int comm_child_read(char **payload, size_t *len) {
	return comm_read(comm[0][0], payload, len);
}

bool comm_child_write(uint8_t type, const char *payload, size_t len) {
	return comm_write(comm[1][1], type, payload, len);
}

int comm_main_read(char **payload, size_t *len) {
	return comm_read(comm[1][0], payload, len);
}

bool comm_main_write(uint8_t type, const char *payload, size_t len) {
	return comm_write(comm[0][1], type, payload, len);
}

int get_comm_reply_fd(void) {
	return comm[1][0];
}

bool write_comm_password(struct swaylock_password *pw) {
	size_t size = pw->len + 1;
	char *copy = password_buffer_create(size);
	if (!copy) {
		clear_password_buffer(pw);
		return false;
	}
	memcpy(copy, pw->buffer, size);
	clear_password_buffer(pw);
	bool ok = comm_main_write(COMM_MSG_PASSWORD, copy, size);
	/* password_buffer_destroy zeros the memory before freeing */
	password_buffer_destroy(copy, size);
	return ok;
}

bool spawn_comm_child(void) {
	if (pipe(comm[0]) != 0) {
		swaylock_log_errno(LOG_ERROR, "failed to create pipe");
		return false;
	}
	if (pipe(comm[1]) != 0) {
		swaylock_log_errno(LOG_ERROR, "failed to create pipe");
		return false;
	}
	pid_t child = fork();
	if (child < 0) {
		swaylock_log_errno(LOG_ERROR, "failed to fork");
		return false;
	} else if (child == 0) {
		struct sigaction sa = {
			.sa_handler = SIG_IGN,
		};
		sigaction(SIGUSR1, &sa, NULL);
		close(comm[0][1]);
		close(comm[1][0]);
		/* Redirect stdin and stdout to /dev/null so the PAM
		 * module cannot fall back to prompting on the terminal
		 * if the main process exits or authd is unavailable. */
		int devnull = open("/dev/null", O_RDWR);
		if (devnull >= 0) {
			dup2(devnull, STDIN_FILENO);
			dup2(devnull, STDOUT_FILENO);
			if (devnull > STDOUT_FILENO) {
				close(devnull);
			}
		}
		run_pw_backend_child();
		/* run_pw_backend_child calls exit(); unreachable */
		abort();
	}
	close(comm[0][0]);
	close(comm[1][1]);
	return true;
}